numpy/doc/structured_arrays.py

```diff
@@ -6,231 +6,430 @@
 
- >>> x[['y','x']]
- array([(2.5, 1.5), (4.0, 3.0), (3.0, 1.0)],
-      dtype=[('y', '<f4'), ('x', '<f4')])
+ >>> y.dtype, y.shape, y.strides
+ (dtype('float32'), (2,), (12,))
 
-Filling structured arrays
-=========================
+Note that unlike other numpy scalars void structured scalars act like views
+into the original ndarray. They support access and assignment by field name::
 
-Structured arrays can be filled by field or row by row. ::
+ >>> x = np.array([(1,2),(3,4)], dtype=[('foo', 'i8'), ('bar', 'f4')])
+ >>> s = x[0]
+ >>> s['bar'] = 100
+ >>> x
+ array([(1, 100.0), (3, 4.0)],
+       dtype=[('foo', '<i8'), ('bar', '<f4')])
+
+In this sense, 1d structured arrays behave similarly to 2d non-structured
+arrays, as indexing using a single index returns a view into the array rather
+than an immutable scalar.
 
- >>> arr = np.zeros((5,), dtype=[('var1','f8'),('var2','f8')])
- >>> arr['var1'] = np.arange(5)
+Accessing Multiple Fields
+```````````````````````````
 
-If you fill it in row by row, it takes a take a tuple
-(but not a list or array!)::
+Numpy allows multi-field indices, where the index is a list of field names::
 
- >>> arr[0] = (10,20)
- >>> arr
- array([(10.0, 20.0), (1.0, 0.0), (2.0, 0.0), (3.0, 0.0), (4.0, 0.0)],
-      dtype=[('var1', '<f8'), ('var2', '<f8')])
+ >>> a = np.zeros(3, dtype=[('a', 'i8'), ('b', 'i4'), ('c', 'f8')])
+ >>> a[['a', 'c']]
+ array([(0, 0.0), (0, 0.0), (0, 0.0)],
+       dtype={'names':['a','c'], 'formats':['<i8','<f8'], 'offsets':[0,11], 'itemsize':19})
+ >>> a[['a', 'c']] = (2, 3)
+ >>> a
+ array([(2, 0, 3.0), (2, 0, 3.0), (2, 0, 3.0)],
+       dtype=[('a', '<i8'), ('b', '<i4'), ('c', '<f8')])
+
+The resulting array is a view of the original array, such that assignment to
+the view modifies the original array. The fields of the view will be in the
+order they were indexed. Note that unlike for single-field indexing here
+the view's dtype has the same itemsize as the original array and has fields at
+the same offsets as in the original array, and unindexed fields are merely
+missing.
+
+Since this view is a structured array itself, it follows the assignment rules
+described above. For example, this means that one can swap the values of two
+fields::
+
+ >>> a[['a', 'c']] = a[['c', 'a']]
+
+Viewing Structured Arrays Containing Objects
+--------------------------------------------
+
+In order to prevent clobbering of object pointers in fields of
+:class:`numpy.object` type
+.. (see :ref:`Object Arrays <arrays.object>`)
+, numpy currently does not allow views of structured arrays containing objects.
+
+Structure Comparison
+--------------------
+
+If the dtypes of two structured arrays are equivalent, testing the equality of
+the arrays will result in a boolean array with the dimension of the original
+arrays, with elements set to True where all fields of the correspnding
+structures are equal. Structured dtypes are equivalent if the field names,
+dtypes and titles are the same, ignoring endianness.
```

Comment:
This is misleading - we should clarify whether "equivalent dtypes" are such that `dt1 == dt2`, or if simply `np.can_cast(dt1, dt2)`